File: src/fos/backend/api/routes/scenarios.py
"""Scenario API routes.

Provides endpoints for fetching scenario definitions and actions.
"""

from litestar import Router, get
from litestar.exceptions import NotFoundException

from fos.core.scenarios import get_all_scenarios, get_scenario, get_scenario_actions
from fos.backend.services.gaworld_agents import get_default_gaworld_agents


@get()
async def list_scenarios() -> list[dict]:
    """Get all scenario definitions.

    Returns list of scenarios with id, name, category, description,
    parameters shape, and actions.
    """
    return get_all_scenarios()


@get("/{scenario_id:str}")
async def get_scenario_detail(scenario_id: str) -> dict:
    """Get full details of a single scenario.

    Args:
        scenario_id: The unique scenario identifier

    Returns:
        Full scenario dict

    Raises:
        NotFoundException: If scenario not found
    """
    scenario = get_scenario(scenario_id)
    if not scenario:
        raise NotFoundException(f"Scenario '{scenario_id}' not found")
    if scenario_id == "gaworld":
        scenario = dict(scenario)
        scenario["default_agents"] = get_default_gaworld_agents()
    return scenario


@get("/{scenario_id:str}/actions")
async def get_scenario_actions_endpoint(scenario_id: str) -> list[dict]:
    """Get actions for a specific scenario.

    Args:
        scenario_id: The unique scenario identifier

    Returns:
        List of action dicts with name and description.
        Returns empty list if scenario not found.
    """
    return get_scenario_actions(scenario_id)


@get("/{scenario_id:str}/default-agents")
async def get_scenario_default_agents(
    scenario_id: str,
    agent_ids: str | None = None,
) -> list[dict]:
    """Get default editable agents for scenarios that provide bundled profiles."""
    if scenario_id != "gaworld":
        raise NotFoundException(f"Scenario '{scenario_id}' has no default agents")

    return get_default_gaworld_agents(agent_ids)


@get("/gaworld/default-agents")
async def get_gaworld_default_agents(agent_ids: str | None = None) -> list[dict]:
    """Get editable bundled GAWorld profile agents."""
    return get_default_gaworld_agents(agent_ids)


@get("/default-agents/gaworld")
async def get_gaworld_default_agents_fallback(agent_ids: str | None = None) -> list[dict]:
    """Get editable bundled GAWorld profile agents using a static fallback path."""
    return get_default_gaworld_agents(agent_ids)


router = Router(
    path="/scenarios",
    route_handlers=[
        list_scenarios,
        get_gaworld_default_agents,
        get_gaworld_default_agents_fallback,
        get_scenario_actions_endpoint,
        get_scenario_default_agents,
        get_scenario_detail,
    ],
)

File: frontend/services/scenarios.ts
/**
 * Scenario service for fetching scenario definitions.
 */

import { getApiBase } from './base';
import type { Agent } from '../types';

export interface ScenarioParam {
  key: string;
  label: string;
  description?: string;
  category?: string;
  type: 'integer' | 'number' | 'float' | 'string' | 'text' | 'boolean' | 'array';
  default: unknown;
  ui_hint?:
    | 'slider'
    | 'number'
    | 'percentage'
    | 'text'
    | 'textarea'
    | 'select'
    | 'toggle'
    | 'list'
    | 'multiselect'
    | 'key_value'
    | 'drag_list'
    | string;
  min?: number;
  max?: number;
  step?: number;
  options?: string[];
  placeholder?: string;
  generates_actions?: boolean;  // Flag indicating this parameter generates actions
}

export interface ActionDef {
  id?: string;
  name: string;
  description: string;
}

export interface ScenarioData {
  id: string;
  name: string;
  category: string;
  description: string;
  interaction_mode?: 'simultaneous' | 'paired' | 'sequential';
  display_type?: 'payoff_matrix' | 'params';
  matrix_meta?: {
    symmetric: boolean;
    rows: string[];
    cols: string[];
    cells: Record<string, string>;
  };
  parameters: ScenarioParam[];
  actions: ActionDef[];
  category_actions?: ActionDef[];
  default_action_ids?: string[];
  default_agents?: Agent[];
}

const API_BASE = getApiBase().replace(/\/+$/, '');

/**
 * Fetch all scenarios.
 */
export async function getAllScenarios(): Promise<ScenarioData[]> {
  const response = await fetch(`${API_BASE}/scenarios`);
  if (!response.ok) {
    throw new Error(`Failed to fetch scenarios: ${response.statusText}`);
  }
  return response.json();
}

/**
 * Fetch a single scenario by ID.
 */
export async function getScenario(id: string): Promise<ScenarioData> {
  const response = await fetch(`${API_BASE}/scenarios/${id}`);
  if (!response.ok) {
    throw new Error(`Failed to fetch scenario: ${response.statusText}`);
  }
  return response.json();
}

/**
 * Fetch actions for a scenario.
 */
export async function getScenarioActions(id: string): Promise<ActionDef[]> {
  const response = await fetch(`${API_BASE}/scenarios/${id}/actions`);
  if (!response.ok) {
    throw new Error(`Failed to fetch scenario actions: ${response.statusText}`);
  }
  return response.json();
}

/**
 * Fetch editable default agents for a scenario that provides bundled profiles.
 */
export async function getScenarioDefaultAgents(id: string, agentIds?: string): Promise<Agent[]> {
  const params = new URLSearchParams();
  if (agentIds?.trim()) {
    params.set('agent_ids', agentIds);
  }
  const suffix = params.toString() ? `?${params.toString()}` : '';
  const response = await fetch(`${API_BASE}/scenarios/${id}/default-agents${suffix}`);
  if (response.ok) {
    return response.json();
  }
  if (id === 'gaworld' && response.status === 404) {
    const fallbackResponse = await fetch(`${API_BASE}/scenarios/default-agents/gaworld${suffix}`);
    if (fallbackResponse.ok) {
      return fallbackResponse.json();
    }
    if (fallbackResponse.status === 404) {
      const scenarioResponse = await fetch(`${API_BASE}/scenarios/${id}`);
      if (scenarioResponse.ok) {
        const scenario = await scenarioResponse.json() as ScenarioData;
        if (Array.isArray(scenario.default_agents)) {
          return filterScenarioDefaultAgents(scenario.default_agents, agentIds);
        }
      }
    }
    throw new Error(`Failed to fetch default agents: ${fallbackResponse.statusText}`);
  }
  if (!response.ok) {
    throw new Error(`Failed to fetch default agents: ${response.statusText}`);
  }
  return response.json();
}

function filterScenarioDefaultAgents(agents: Agent[], agentIds?: string): Agent[] {
  const ids = agentIds
    ?.split(',')
    .map((id) => id.trim())
    .filter(Boolean);
  if (!ids || ids.length === 0) {
    return agents;
  }
  const selected = new Set(ids);
  return agents.filter((agent) => selected.has(String(agent.id)));
}
